Escape characters are handled incorrectly in inline config comments

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a change: https://eslint.org/docs/developer-guide/contributing/changes
    3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** v4.7.2
* **Node Version:** v8.5.0
* **npm Version:** Yarn v1.0.2

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```yml
rules:
    id-match: [2, "^(([^$\\W]|\\$[a-f\\d]{2})+|[$_]\\w*|[^\\W\\d]\\w*|[A-Z]([A-Z_]*[A-Z])?)$", {
        properties: true
    }]
```

</details>

**What did you do? Please include the actual source code causing the issue.**

<!-- Paste the source code below: -->
```js
/* eslint id-match: [2, "^(([^$\\W]|\\$[a-f\\d]{2})+|[$_]\\w*|[^\\W\\d]\\w*|[A-Z]([A-Z_]*[A-Z])?)$", {properties: true}] */

function is$2dvoid(value) {
    return value == null
}
```

**What did you expect to happen?** It to pass (as it does without the comment)

**What actually happened? Please include the actual, raw output from ESLint.**

```
/path/to/file.js
  3:10  error  Identifier 'is$2dvoid' does not match the pattern '^(([^$\\W]|\\$[a-f\\d]{2})+|[$_]\\w*|[^\\W\\d]\\w*|[A-Z]([A-Z_]*[A-Z])?)$'  id-match

✖ 1 problem (1 error, 0 warnings)
```

--- a/package.json
+++ b/package.json
@@ -69,7 +69,7 @@
     "is-glob": "^4.0.0",
     "js-yaml": "^3.13.1",
     "json-stable-stringify-without-jsonify": "^1.0.1",
-    "levn": "^0.3.0",
+    "levn": "^0.4.1",
     "lodash": "^4.17.14",
     "minimatch": "^3.0.4",
     "natural-compare": "^1.4.0",

--- a/package.json
+++ b/package.json
@@ -73,7 +73,7 @@
     "lodash": "^4.17.14",
     "minimatch": "^3.0.4",
     "natural-compare": "^1.4.0",
-    "optionator": "^0.8.3",
+    "optionator": "^0.9.1",
     "progress": "^2.0.0",
     "regexpp": "^3.0.0",
     "semver": "^7.1.1",

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -2734,7 +2734,7 @@ describe("Linter", () => {
 
     describe("when evaluating code with comments which have colon in its value", () => {
         const code = String.raw`
-/* eslint max-len: [2, 100, 2, {ignoreUrls: true, ignorePattern: "data:image\/|\\s*require\\s*\\(|^\\s*loader\\.lazy|-\\*-"}] */
+/* eslint max-len: [2, 100, 2, {ignoreUrls: true, ignorePattern: "data:image\\/|\\s*require\\s*\\(|^\\s*loader\\.lazy|-\\*-"}] */
 alert('test');
 `;
 
@@ -2743,29 +2743,35 @@ alert('test');
 
             assert.strictEqual(messages.length, 1);
             assert.strictEqual(messages[0].ruleId, "max-len");
-            assert.strictEqual(messages[0].message, "This line has a length of 128. Maximum allowed is 100.");
+            assert.strictEqual(messages[0].message, "This line has a length of 129. Maximum allowed is 100.");
             assert.include(messages[0].nodeType, "Program");
         });
     });
 
     describe("when evaluating code with comments that contain escape sequences", () => {
         const code = String.raw`
-/* eslint max-len: ["error", 1, { ignoreComments: true, ignorePattern: "console\.log\\(" }] */
+/* eslint max-len: ["error", 1, { ignoreComments: true, ignorePattern: "console\\.log\\(" }] */
 console.log("test");
+consolexlog("test2");
 var a = "test2";
 `;
 
         it("should validate correctly", () => {
             const config = { rules: {} };
-
             const messages = linter.verify(code, config, filename);
+            const [message1, message2] = messages;
 
-            assert.strictEqual(messages.length, 1);
-            assert.strictEqual(messages[0].ruleId, "max-len");
-            assert.strictEqual(messages[0].message, "This line has a length of 16. Maximum allowed is 1.");
-            assert.strictEqual(messages[0].line, 4);
-            assert.strictEqual(messages[0].column, 1);
-            assert.include(messages[0].nodeType, "Program");
+            assert.strictEqual(messages.length, 2);
+            assert.strictEqual(message1.ruleId, "max-len");
+            assert.strictEqual(message1.message, "This line has a length of 21. Maximum allowed is 1.");
+            assert.strictEqual(message1.line, 4);
+            assert.strictEqual(message1.column, 1);
+            assert.include(message1.nodeType, "Program");
+            assert.strictEqual(message2.ruleId, "max-len");
+            assert.strictEqual(message2.message, "This line has a length of 16. Maximum allowed is 1.");
+            assert.strictEqual(message2.line, 5);
+            assert.strictEqual(message2.column, 1);
+            assert.include(message2.nodeType, "Program");
         });
     });
 

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -2733,14 +2733,31 @@ describe("Linter", () => {
     });
 
     describe("when evaluating code with comments which have colon in its value", () => {
-        const code = "/* eslint max-len: [2, 100, 2, {ignoreUrls: true, ignorePattern: \"data:image\\/|\\s*require\\s*\\(|^\\s*loader\\.lazy|-\\*-\"}] */\nalert('test');";
+        const code = "/* eslint max-len: [2, 100, 2, {ignoreUrls: true, ignorePattern: \"data:image\\/|\\\\s*require\\\\s*\\\\(|^\\\\s*loader\\\\.lazy|-\\\\*-\"}] */\nalert('test');";
 
         it("should not parse errors, should report a violation", () => {
             const messages = linter.verify(code, {}, filename);
 
             assert.strictEqual(messages.length, 1);
             assert.strictEqual(messages[0].ruleId, "max-len");
-            assert.strictEqual(messages[0].message, "This line has a length of 122. Maximum allowed is 100.");
+            assert.strictEqual(messages[0].message, "This line has a length of 128. Maximum allowed is 100.");
+            assert.include(messages[0].nodeType, "Program");
+        });
+    });
+
+    describe("when evaluating code with comments that contain escape sequences", () => {
+        const code = '/* eslint max-len: ["error", 1, { ignoreComments: true, ignorePattern: "console\\.log\\\\(" }] */\nconsole.log("test");\nvar a = "test2";';
+
+        it("should validate correctly", () => {
+            const config = { rules: {} };
+
+            const messages = linter.verify(code, config, filename);
+
+            assert.strictEqual(messages.length, 1);
+            assert.strictEqual(messages[0].ruleId, "max-len");
+            assert.strictEqual(messages[0].message, "This line has a length of 16. Maximum allowed is 1.");
+            assert.strictEqual(messages[0].line, 3);
+            assert.strictEqual(messages[0].column, 1);
             assert.include(messages[0].nodeType, "Program");
         });
     });

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -2733,7 +2733,10 @@ describe("Linter", () => {
     });
 
     describe("when evaluating code with comments which have colon in its value", () => {
-        const code = "/* eslint max-len: [2, 100, 2, {ignoreUrls: true, ignorePattern: \"data:image\\/|\\\\s*require\\\\s*\\\\(|^\\\\s*loader\\\\.lazy|-\\\\*-\"}] */\nalert('test');";
+        const code = String.raw`
+/* eslint max-len: [2, 100, 2, {ignoreUrls: true, ignorePattern: "data:image\/|\\s*require\\s*\\(|^\\s*loader\\.lazy|-\\*-"}] */
+alert('test');
+`;
 
         it("should not parse errors, should report a violation", () => {
             const messages = linter.verify(code, {}, filename);
@@ -2746,7 +2749,11 @@ describe("Linter", () => {
     });
 
     describe("when evaluating code with comments that contain escape sequences", () => {
-        const code = '/* eslint max-len: ["error", 1, { ignoreComments: true, ignorePattern: "console\\.log\\\\(" }] */\nconsole.log("test");\nvar a = "test2";';
+        const code = String.raw`
+/* eslint max-len: ["error", 1, { ignoreComments: true, ignorePattern: "console\.log\\(" }] */
+console.log("test");
+var a = "test2";
+`;
 
         it("should validate correctly", () => {
             const config = { rules: {} };
@@ -2756,7 +2763,7 @@ describe("Linter", () => {
             assert.strictEqual(messages.length, 1);
             assert.strictEqual(messages[0].ruleId, "max-len");
             assert.strictEqual(messages[0].message, "This line has a length of 16. Maximum allowed is 1.");
-            assert.strictEqual(messages[0].line, 3);
+            assert.strictEqual(messages[0].line, 4);
             assert.strictEqual(messages[0].column, 1);
             assert.include(messages[0].nodeType, "Program");
         });

